### ISSUE TITLE
Excessive memory usage when using fit_generator()

### ISSUE THREAD

**Issue by Punyaslok** [keras-team/keras#11671]
`fit_generator()` seems to keep consuming memory throughout an epoch if objects defined inside the generator class are not converted into a numpy array. RAM keeps on getting filled until an epoch ends. For large enough datasets, this leads to RAM completely filling up before a single epoch is completed.

Is it possible that the objects keep getting duplicated?

Python = 3.6.6
Keras = 2.2.4
Tensorflow = 1.12.0

[Here is a sample script to reproduce the issue.](https://gist.github.com/Punyaslok/5a3dd01bd004ba84ae37329799d47430)

**Comment by Dref360** [comment]
The Sequence object is duplicated `workers` time. Also, if you are on UNIX, by default the process is forked which consumes more memory than using `spawn`. 

I think that when you use np array, only the pointer is duplicated? If I'm right, this is not process-safe.

**Comment by sziraqui** [comment]
Same problem. I can load 1184 large np arrays using fit() but only 16 such arrays using fit_generator before memory hits 7.6GB and crashes. Is there a workaround?

**Comment by Dref360** [comment]
@sziraqui  it depends on your fit_generator call.

Let's say you have (workers=4, max_queue_size=10)

Then, you load 16 arrays * workers * max_queue_size * 2 (train and validation)

It grows rather quickly.

**Comment by sziraqui** [comment]
I have workers=1, max_queue_size=2
My generator function yields X, which has 32 numpy arrays of shape 15,50,100,3 and Y which has 32 arrays of shape 15,28. I m pre-allocating both arrays and reusing them for every batch.
Using fit, I can load 1184 samples and train them without a crashing. But with fit_generator, processing 512 samples in batches of 32 or even 16 fills memory after 4-5 epochs. Before posting previous comment, I was not pre-allocating X,Y and the situation was even worse. But now memory consumption increases less rapidly but still is higher than loading whole sample at once.

**Comment by sziraqui** [comment]
I have observed the effect to depend on `sample_size/batch_size` instead of only batch_size.

**Comment by y-bai** [comment]
I found the issue related to the numpy version.

if I used keras 2.2.4, tf 1.12.0, python 3.6.5 and numpy>=1.16,0 then I encountered the same issue.
but when I downgraded the numpy version from 1.16.0 to 1.14.3, with keras 2.2.4, tf 1.12.0, python 3.6.5. the problem solved.

The problem was also solved when using  python 3.6.8, numpy==1.16.4
tensorflow 1.12, keras 2.2.4.

Thus, I believe the problem is related with the version between python and numpy.

@Dref360
